src/Data/GADT/Coerce.hs

```diff
@@ -0,0 +1,107 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE PolyKinds           #-}
+#if __GLASGOW_HASKELL__ >= 810
+{-# LANGUAGE StandaloneKindSignatures #-}
+#endif
+module Data.GADT.Coerce (
+    GCoercible (..),
+    defaultGcoercible,
+    ) where
+
+import Data.Functor.Product (Product (..))
+import Data.Functor.Sum     (Sum (..))
+import Data.IORef           (IORef)
+import Data.STRef           (STRef)
+import Data.Type.Equality   ((:~:) (..))
+import Data.Type.Coercion
+
+import Unsafe.Coerce        (unsafeCoerce)
+
+import Data.GADT.Compare    (GEq, geq)
+
+#if MIN_VERSION_base(4,10,0)
+import           Data.Type.Equality ((:~~:) (..))
+import qualified Type.Reflection    as TR
+#endif
+
+#if __GLASGOW_HASKELL__ >= 810
+import Data.Kind (Type, Constraint)
+#endif
+
+-- |A class for type-contexts which contain enough information
+-- to (at least in some cases) decide the coercibility of types
+-- occurring within them.
+#if __GLASGOW_HASKELL__ >= 810
+type GCoercible :: (k -> Type) -> Constraint
+#endif
+class GCoercible f where
+    gcoercible :: f a -> f b -> Maybe (Coercion a b)
+
+-- |If 'f' has a 'GEq' instance, this function makes a suitable default
+-- implementation of 'gcoercible'.
+defaultGcoercible :: GEq f => f a -> f b -> Maybe (Coercion a b)
+defaultGcoercible x y = fmap repr $ geq x y
+
+instance GCoercible ((:~:) a) where
+    gcoercible = defaultGcoercible
+
+#if MIN_VERSION_base(4,10,0)
+instance GCoercible ((:~~:) a) where
+    gcoercible = defaultGcoercible
+
+instance GCoercible TR.TypeRep where
+    gcoercible = defaultGcoercible
+#endif
+
+instance (GCoercible a, GCoercible b) => GCoercible (Sum a b) where
+    gcoercible (InL x) (InL y) = gcoercible x y
+    gcoercible (InR x) (InR y) = gcoercible x y
+    gcoercible _ _ = Nothing
+
+instance (GCoercible a, GCoercible b) => GCoercible (Product a b) where
```

Comment:
This is weird (also in existing `GEq` instance. Either side would be enough, won't it?

Reply:
Huh? It is possible for one to fail and the other to succeed, is it not?

Reply:
For `TestEquality` yes only one is sufficient.

Reply:
Yes, I was thinking of `TestEquality`, and it doesn't seem to have instance for `Product` so all is fine.

(btw, that's another point not to have `TestEquality` as a super-class for `GEq`).